contracts/TopShotMarketV3.cdc

```diff
@@ -0,0 +1,345 @@
+
+                    // borrow a reference to the main top shot collection
+                    let ownerCollectionRef = self.ownerCollection.borrow()
+                        ?? panic("Could not borrow owner collection reference")
+
+                    // deposit the withdrawn moment into the main collection
+                    ownerCollectionRef.deposit(token: <-token)
+                } else {
+                    panic("Token with the specified ID is not already for sale in either sale collection")
+                }
+
+            } else {
+                panic("Token with the specified ID is not already for sale in either sale collection")
+            }
+        }
+
+        /// purchase lets a user send tokens to purchase an NFT that is for sale
+        /// the purchased NFT is returned to the transaction context that called it
+        ///
+        /// Parameters: tokenID: the ID of the NFT to purchase
+        ///             buyTokens: the fungible tokens that are used to buy the NFT
+        ///
+        /// Returns: @TopShot.NFT: the purchased NFT
+        pub fun purchase(tokenID: UInt64, buyTokens: @FungibleToken.Vault): @TopShot.NFT {
+
+            // First check this sale collection for the NFT
+            if self.prices[tokenID] != nil {
+                assert(
+                    buyTokens.balance == self.prices[tokenID]!,
+                    message: "Not enough tokens to buy the NFT!"
+                )
+
+                // Read the price for the token
+                let price = self.prices[tokenID]!
+
+                // Set the price for the token to nil
+                self.prices[tokenID] = nil
+
+                // Take the cut of the tokens that the beneficiary gets from the sent tokens
+                let beneficiaryCut <- buyTokens.withdraw(amount: price*self.cutPercentage)
+
+                // Deposit it into the beneficiary's Vault
+                self.beneficiaryCapability.borrow()!
+                    .deposit(from: <-beneficiaryCut)
+                
+                // Deposit the remaining tokens into the owners vault
+                self.ownerCapability.borrow()!
+                    .deposit(from: <-buyTokens)
+
+                emit MomentPurchased(id: tokenID, price: price, seller: self.owner?.address)
+
+                // Return the purchased token
+                let boughtMoment <- self.ownerCollection.borrow()!.withdraw(withdrawID: tokenID) as! @TopShot.NFT
+
+                return <-boughtMoment
+
+            // If not found in this SaleCollection, check V1
+            } else if let v1Market = self.marketV1Capability {
+                let v1MarketRef = v1Market.borrow()!
+
+                return <-v1MarketRef.purchase(tokenID: tokenID, buyTokens: <-buyTokens)
+            }
+            
+            panic("No token matching this ID for sale!")
+        }
+
+        /// changePercentage changes the cut percentage of the tokens that are for sale
```

Comment:
Yeah, the cutPercentage is taken at purchase time, so if someone changes the percentage, it will apply to all the tokens in the sale. I thought we had decided that that was what we wanted. Is it not desired any more?

Reply:
I don't remember why we set it this way though, do you?

It feels odd that you can change the cutPercentage of EXISTING listed items.

A malicious operator could wait for everyone to list and say the cutPercentage is 1% but after everything is listed they can set it to 50% or something?

Reply:
maybe add a check in purchase to validate cut percentage?

Reply:
what exactly does that achieve thou?
buyer: they pay the same amount regardless of cut percentage
seller: they would be the party that authorize the cut percentage change?

Reply:
@EricLin2004 @rrrkren @joshuahannan clarification: do we want to remove `changePercentage` entirely?

Reply:
Yeah, the seller is the one who sets the cut percentage, so it would only change it for their sales, which would only affect them. The buyer still pays the same price. I think we should keep the `changePercentage` function in case the top shot team decides to change it in the future

Reply:
Just chatted with @EricLin2004 - he would like the `changePercentage` to be removed as we don't currently use it and likely will not (can't think of any reason that it would make sense for a seller to be able to change the cut percentage) and since the contracts are public we don't want to set a precedent for it that others may use as an example

Reply:
Removed `changePercentage`